pallets/pips/src/lib.rs

```diff
@@ -809,194 +864,264 @@ decl_module! {
+        /// * `IncorrectProposalState` if the proposal isn't pending.
+        /// * `ProposalOnCoolOffPeriod` if the proposal is cooling off.
+        /// * `NotByCommittee` if the proposal isn't by a committee.
+        #[weight = (1_000_000_000, DispatchClass::Operational, Pays::Yes)]
+        pub fn approve_committee_proposal(origin, id: PipId) {
+            // 1. Only GC can do this.
             T::VotingMajorityOrigin::ensure_origin(origin)?;
-            ensure!(<Proposals<T>>::contains_key(id), Error::<T>::NoSuchProposal);
-            // Check that the proposal is pending
+
+            // 2. Proposal must be pending.
             Self::is_proposal_state(id, ProposalState::Pending)?;
-            Self::refund_proposal(id);
-            Self::update_proposal_state(id, ProposalState::Killed);
-            Self::prune_data(id, Self::prune_historical_pips());
+
+            // 3. Proposal must not be cooling-off and must be by committee.
+            let meta = Self::proposal_metadata(id).ok_or_else(|| Error::<T>::NoSuchProposal)?;
+            let curr_block_number = <system::Module<T>>::block_number();
+            ensure!(meta.cool_off_until <= curr_block_number, Error::<T>::ProposalOnCoolOffPeriod);
+            ensure!(matches!(meta.proposer, Proposer::Committee(_)), Error::<T>::NotByCommittee);
+
+            // 4. All is good, schedule PIP for execution.
+            let current_did = Context::current_identity::<Identity<T>>().unwrap_or_default();
+            Self::schedule_pip_for_execution(current_did, id);
         }
 
-        /// An emergency stop measure to kill a proposal. Governance committee can kill
-        /// a proposal at any time.
+        /// Rejects the PIP given by the `id`, refunding any bonded funds,
+        /// assuming it hasn't been cancelled or executed.
+        /// Note that cooling-off and proposals scheduled-for-execution can also be rejected.
+        ///
+        /// # Errors
+        /// * `BadOrigin` unless a GC voting majority executes this function.
+        /// * `NoSuchProposal` if the PIP with `id` doesn't exist.
+        /// * `IncorrectProposalState` if the proposal was cancelled or executed.
+        #[weight = (550_000_000, DispatchClass::Operational, Pays::Yes)]
+        pub fn reject_proposal(origin, id: PipId) {
+            T::VotingMajorityOrigin::ensure_origin(origin)?;
+            let proposal = Self::proposals(id).ok_or_else(|| Error::<T>::NoSuchProposal)?;
+            ensure!(
+                matches!(proposal.state, ProposalState::Pending | ProposalState::Scheduled),
+                Error::<T>::IncorrectProposalState,
+            );
+            Self::maybe_unschedule_pip(id, proposal.state);
+            Self::maybe_unsnapshot_pip(id, proposal.state);
+            Self::unsafe_reject_proposal(id);
+        }
+
+        /// Prune the PIP given by the `id`, refunding any funds not already refunded.
+        /// *No restrictions* on the PIP applies.
+        ///
```

Comment:
I think we should only allow a PIP to be pruned if it is "completed" (i.e. Cancelled, Rejected, Failed or Executed). i.e. no further action can be taken on the PIP, so we can reclaim storage.